src/engagement_db_coda_sync/lib.py

```diff
@@ -1,15 +1,102 @@
+    :type coda_config: src.engagement_db_coda_sync.configuration.CodaSyncConfiguration
+    :param google_cloud_credentials_file_path: Path to a Google Cloud service account credentials file 
+                                               to use to access the credentials bucket.
+    :type google_cloud_credentials_file_path: str
+    """
+    all_datasets_have_user_file_url = all(
+        dataset_config.dataset_users_file_url is not None for dataset_config in coda_config.dataset_configurations)
+
+    default_config_user_ids = []
+    if not all_datasets_have_user_file_url:
+        assert coda_config.project_users_file_url is not None, \
+         f"Specify user ids for coda datasets in CodaDatasetConfiguration or user ids for this project in CodaSyncConfiguration"
+        default_config_user_ids = get_coda_users_from_gcloud(coda_config.project_users_file_url, google_cloud_credentials_file_path)
+
+    ws_correct_dataset_code_scheme = coda_config.ws_correct_dataset_code_scheme
+    for dataset_config in coda_config.dataset_configurations:
+        config_user_ids = []
+        if dataset_config.dataset_users_file_url:
+            config_user_ids = get_coda_users_from_gcloud(dataset_config.dataset_users_file_url, google_cloud_credentials_file_path)
+        else:
+            config_user_ids = default_config_user_ids
+
+        coda_user_ids = coda.get_dataset_user_ids(dataset_config.coda_dataset_id)
+        if coda_user_ids:
+            config_user_ids = list(set(config_user_ids) - set(coda_user_ids))
+            coda.update_dataset_user_ids(dataset_config.coda_dataset_id, config_user_ids)
+        else:
+            coda.set_dataset_user_ids(dataset_config.coda_dataset_id, config_user_ids)
+
+        repo_code_schemes = []
+        for code_scheme_config in dataset_config.code_scheme_configurations:
+            for count in range(1, code_scheme_config.coda_code_schemes_count + 1):
+                if count == 1:
+                    repo_code_schemes.append(code_scheme_config.code_scheme)
+                else:
+                    code_scheme_copy = code_scheme_config.code_scheme.copy()
+                    code_scheme_copy.scheme_id = f"{code_scheme_copy.scheme_id}-{count}"
+                    repo_code_schemes.append(code_scheme_copy)
+        repo_code_schemes.append(ws_correct_dataset_code_scheme)
+        repo_code_schemes_lut = {code_scheme.scheme_id: code_scheme for code_scheme in repo_code_schemes}
+
+        coda_code_schemes = coda.get_all_code_schemes(dataset_config.coda_dataset_id)
+        coda_scheme_ids = [code_scheme.scheme_id for code_scheme in coda_code_schemes]
+
+        for coda_scheme_id in coda_scheme_ids:
+            if coda_scheme_id not in repo_code_schemes_lut.keys():
+                log.warning(f"There are code schemes in coda not in this repo; The code schemes will be ignored")
+                coda_code_schemes.remove(repo_code_schemes_lut[coda_scheme_id])
```

Comment:
Should this be
```suggestion
                coda_code_schemes.remove(coda_scheme_id)
```
?

Reply:
Not really, we should remove the code scheme with the above scheme id, I have updated the code block thanks.